### PR TITLE
[TLS 1.3] Basic Server Implementation

### PR DESCRIPTION
**Pull request by reneme**
## Pull Request Dependencies

* ~~First review/merge https://github.com/randombit/botan/pull/3062 and rebase.~~ That'll make this much easier digestable, I hope.

## Description

This provides a basic TLS 1.3 server implementation with the most common functionality. Relevant BoGo tests pass already, but many more are still disabled. The ordinary handshake (with or without Hello Retry Request) and subsequent application data exchange works. Also a protocol version downgrade to TLS 1.2 works fine.

For the sake of reviewability, I am planning to stabilize the current feature set and aim for review/merge to master. Further functionality will come. See below.

### New Callback: `bool tls_peer_closed_connection()`

This is called when we received a "close_notify" from the peer. In contrast to older revisions, TLS 1.3 does not mandate an immediate in-kind response in that case. Applications can override this callback to get notified that no more data is to be expected from the peer. Though they could legally continue sending data and finish with `Channel::close()` at their own discretion.

Returning `true` from this callback will make TLS 1.3 behave just like 1.2 and reply with a "close_notify" immediately. This is the default behaviour.

## Missing Functionality

* [ ] PSK Mode
* [ ] Resumption
* [ ] Creation and sending of Session Tickets
* [ ] Sending or receiving of Early Data
* [x] Client Authentication (#3081)
* [x] OCSP Stapling (#3112)
* [x] ALPN (#3117 )
* probably more...

### CODE REVIEW

**Comment by reneme** [comment]
Milestone: This can now successfully handshake and exchange messages with `openssl`:

#### Terminal 1

```
./botan tls_server src/tests/data/tls_13_rfc8448/server_certificate.pem src/tests/data/tls_13_rfc8448/server_key.pem --port=8443 --policy=src/tests/data/tls-policy/rfc8448_1rtt.txt
```

#### Terminal 2

```
openssl s_client localhost:8443
```

Next steps: Support for Hello Retry Request and integration with basic BoGo tests. For the latter, I hope we can find a few server tests that don't require resumption support.

**Comment by reneme** [comment]
I'll clean up the commit history tomorrow and would consider this ready for review/merge now.

**Review by lieser — COMMENTED** [review]
Incomplete review of old state, so some comments may already be outdated.

File: src/lib/tls/msg_server_hello.cpp
Comment (reneme) [reply src/lib/tls/msg_server_hello.cpp:795]: Yes, albeit somewhat implicit, when processing the Client Hello in `tls_server_impl_13.cpp`. For now we must just assume that both extensions are availabe, because we don't support any PSK-mode in the server. Once PSK-based negotiation (particularly without ephemeral key exchange) is being implemented we will need to re-iterate this.

File: src/lib/tls/tls13/tls_server_impl_13.cpp
Comment (reneme) [reply src/lib/tls/tls13/tls_server_impl_13.cpp:218]: On a second note: Probably we should adapt the `Extensions.get<...>()` API to be less risky. But that's something I would do out-of-band.

**Comment by randombit** [comment]
I'm deferring a review on this until #3062 lands

**Comment by reneme** [comment]
> I'm deferring a review on this until #3062 lands

Rebased to master. This has no more underlying PR dependencies.

**Comment by randombit** [comment]
I will review this asap

File: src/lib/tls/tls_extensions.h
Comment (reneme) [reply src/lib/tls/tls_extensions.h:731]: Yep, thanks. That was an oversight. I guess, I initially meant to remove it from a multi-parameter constructor. Fixed.

File: src/lib/tls/tls13/tls_server_impl_13.cpp
Comment (reneme) [reply src/lib/tls/tls13/tls_server_impl_13.cpp:192]: Frankly, I don't remember that we discussed that before. Anyway, this might actually point to an implementation issue. So far, the second Client Hello is not compared to the original one. I.e. the server _does not_ ensure that the Client Hello changed only in the aspects outlined by [RFC 8446 4.1.2](https://www.rfc-editor.org/rfc/rfc8446#section-4.1.2).

Curiously, BoGo tests don't seem to cover this. I'll need to look into it. Good catch!

File: src/lib/tls/tls_server.h
Comment (reneme) [reply src/lib/tls/tls_server.h:98]: Indeed, this is a left-over of our previous discussion about session resumption. Removing.

File: src/lib/tls/tls13/tls_server_impl_13.cpp
Comment (reneme) [reply src/lib/tls/tls13/tls_server_impl_13.cpp:192]: https://github.com/randombit/botan/pull/3053/commits/ca25dda541ebcbae61d902aeea41f992630ca3df contains additional validation of the updated Client Hello. It does not specifically validate the contents of extensions, though. I'm frankly not sure whether that is even needed.

Regarding the 'supported version' changing after the Hello Retry Request: If the object is a `Client_Hello_13`, it's highest supported version is definitely TLS 1.3 (otherwise the parser would have classified it as `Client_Hello_12` and we would have run into the check added in the same commit: https://github.com/randombit/botan/pull/3053/commits/ca25dda541ebcbae61d902aeea41f992630ca3df#diff-fcf2471ec5c28b66eae46f45d0e53de60b2afa2fe6f94d219038bc58c9d635be.

**Review by lieser — APPROVED** [review]

**Comment by reneme** [comment]
I left this open despite @lieser approving since you had started a review as well, @randombit.

In case you want to continue reviewing this after your christmas break, please don't hesitate to point that out.

Otherwise, I'd like to merge this soonish and integrate the remaining (approved) PRs. This would particularly simplify development in [the resumption PR](https://github.com/randombit/botan/pull/3140) as it will contain some cross-cutting concerns.

File: src/lib/tls/msg_client_hello.cpp
Comment (randombit) [inline src/lib/tls/msg_client_hello.cpp:1071]: Should we ignore (or reject?) DTLS versions being mixed in here? Or does that happen elsewhere already.

Perhaps for RFC 8446 purposes DTLS counts as an "unknown version" of TLS but that doesn't really seem correct to me.

File: src/lib/tls/msg_server_hello.cpp
Comment (randombit) [inline src/lib/tls/msg_server_hello.cpp:34]: Why this change? Fine with it just curious why it was required.

File: src/lib/tls/tls13/tls_extensions_key_share.cpp
Comment (randombit) [inline src/lib/tls/tls13/tls_extensions_key_share.cpp:266]: Can you open an issue about this? It is definitely a weak point with regards to robustness, especially in extension parsing.

**Review by randombit — APPROVED** [review]
Looks good to me. A few minor questions/comments, approving now to unblock.

File: src/lib/tls/tls13/tls_extensions_key_share.cpp
Comment (reneme) [reply src/lib/tls/tls13/tls_extensions_key_share.cpp:361]: Fair. Now that we have C++20 and ranges I thought that's a good opportunity to use it. For reference:

```C++
return m_client_shares
   | std::views::transform([](const auto& s){ return s.group(); })
   | to<std::vector<Named_Group>>();
```

Caveat: the `to<>()` is actually [proposed for C++23](https://wg21.link/p1206) only. We'd need to implement it ourselves; which seems like a reasonable thing to do for the time being. Sigh!
Godbolt: https://godbolt.org/z/771M7PEzK

But certainly not in this PR.

I guess the modern thing to do, is to return the `view` directly and not create a new owning container at all. These things should happen at least somewhat consistently, though.

File: src/lib/tls/msg_server_hello.cpp
Comment (reneme) [reply src/lib/tls/msg_server_hello.cpp:34]: The c'tor of `Server_Hello_Internal` takes the random as a `std::vector` (which is usually provided from `make_hello_random()`, that returns a `std::vector`). Leaving this as a `std::array` would require a conversion or a c'tor overload just for that.

```C++

std::make_unique<Server_Hello_Internal>(.... , HELLO_RETRY_REQUEST_MARKER, ....);
// vs.
std::make_unique<Server_Hello_Internal>(.... , make_hello_random(), ....);
```

Or use a `std::span` in the `Server_Hello_Internal` c'tor? But wouldn't that cause UB because the `std::span` would refer to data in the temporary `std::vector` falling out of `make_hello_random()`?

My understanding is, that the temporary gets destroyed at the end of the expression (i.e. after the invoked c'tor returned). So as long as the c'tor copies the data out of the span, there should be no lifetime issues. [This quick test](https://godbolt.org/z/jTxKoW6cb) does support my view. Side note: An interesting article about views, ranges, life times and dangling pointers: https://tristanbrindle.com/posts/rvalue-ranges-and-views

But it would result in an unnecessary copy (from the temporary, via the span, into a member of `Server_Hello_Internal`). I think, my vote is to keep the `std::vector<>` for `HELLO_RETRY_REQUEST_MARKER`.

File: src/lib/tls/tls13/tls_extensions_key_share.cpp
Comment (reneme) [reply src/lib/tls/tls13/tls_extensions_key_share.cpp:266]: I suppose [this](https://github.com/randombit/botan/pull/3144) obsoletes the need for an issue.

File: src/lib/tls/msg_client_hello.cpp
Comment (reneme) [reply src/lib/tls/msg_client_hello.cpp:1071]: DTLS as such is not explicitly filtered out anywhere. TLS 1.2 / 1.3 distinctions are typically done using `Protocol_Version::is_pre_tls_13()` or `::is_tls_13_or_later()` which transparently takes DTLS into account.

That said, DTLS 1.3 won't pass the default implementation of `Policy::acceptable_protocol_version()` and won't make it into the result vector. For reference:

```C++
bool Policy::acceptable_protocol_version(Protocol_Version version) const
   {
#if defined(BOTAN_HAS_TLS_13)
   if(version == Protocol_Version::TLS_V13 && allow_tls13())
      return true;
#endif

#if defined(BOTAN_HAS_TLS_12)
   if(version == Protocol_Version::TLS_V12 && allow_tls12())
      return true;

   if(version == Protocol_Version::DTLS_V12 && allow_dtls12())
      return true;
#endif

   return false;
   }
```

File: src/lib/tls/tls13/tls_extensions_key_share.cpp
Comment (reneme) [reply src/lib/tls/tls13/tls_extensions_key_share.cpp:266]: There are some bullet points (under "Potential Future Work") for internal improvements in [the original TLS client PR](https://github.com/randombit/botan/pull/2922) which is also referenced in [the TLS 1.3 "current state" issue](https://github.com/randombit/botan/issues/2990). Admittedly, that's somewhat hidden. Though, I figured a bullet point list would be better than cluttering the Issues with tickets at this point.

**Pull request merged**
